crates/c-api/include/wasmtime/store.h

```diff
@@ -243,6 +243,30 @@ WASM_API_EXTERN void wasmtime_context_set_epoch_deadline(wasmtime_context_t *con
  */
 WASM_API_EXTERN void wasmtime_store_epoch_deadline_callback(wasmtime_store_t *store, wasmtime_error_t* (*func)(wasmtime_context_t*, void*, uint64_t*), void *data);
 
+/**
+ * \brief Insert a file descriptor into the underlying WASI context. 
+ * 
+ * This is used to dynamically bind a file descriptor from the host to a file 
+ * descriptor in an already-built WASI context to make it available to WASI APIs.
+ * 
+ * The `guest_fd` argument is the number of the file descriptor by which it will
+ * be known in WASM and the `host_fd` is the number of the file descriptor on
+ * the host.
+ */
+WASM_API_EXTERN void wasmtime_context_insert_file(wasmtime_context_t *context, uint32_t guest_fd, uint32_t host_fd, uint32_t access_mode);
```

Comment:
Also, could this document the `access_mode` values? I don't believe that maps to octal permissions like `0o777` as one might otherwise expect.

Reply:
This is basically the same as whatever in Rust.

Reply:
> I think that this and the below function will need to use a different type for host_fd on Windows where it'll use int on Unix and HANDLE on Windows.

Would you give an example or directly push to my branch? I do not have a Windows setup for testing any change specific to Windows.

Reply:
> it'll use int on Unix

Seems even 'int' is not the case to me. It seems to be i32 on Unix. 

Opt to use u32 here simply for the type consistency across all fds -- iirc the returned guest_fd is in u32. I can double check.

Reply:
And IMO, it is Rust's fault to select i32 for file descriptors. An apparent better and more consistent choice is uintptr. 

I'm personally against exposing host_fd as i32, but if you insist it should guarantee truthfulness (host_fd: i32, guest_fd: u32) rather than consistency, I can make the corresponding change.

Reply:
I looked again and found Go will only return `uintptr` for the fd. 

I would propose the change to always use `* c_void`(Rust)/`void *` for all guest/host fd for the interface, and leaving the type conversion inside Rust implementation. What do you think? @alexcrichton

Reply:
Sorry I don't have detailed guides and such of how to build on Windows, I also don't develop on Windows myself. This problem must be fixed to merge this PR, however, because I believe your changes will not compile on Windows due to the usage of `std::os::fd` which does not on Windows. CI is passing because only a subset of tests are run for PRs, but the full merge will run a full suite of tests. If you'd like you can include a commit with `prtest:full` somewhere in the commit message to run full CI on this PR.

As for types, IMO this should match what happens in the native platform. This means that `int` should be used in the header file for Unix and `HANDLE` should be used in the header file for Windows. The corresponding Rust types are `std::os::fd::RawFd` and `std::os::windows::io::RawHandle`.

I don't have a better idea of how to define this function other than to just have it defined differently on different platforms.

Reply:
Luckily it worked after some reverse engineering of the CI script of wasmtime and wasmtime-go, now I can finally compile and test on Windows. 

Just pushed a new change to use `void*`/`*mut c_void`/`unsafe.Pointer()` for `host_fd`. How does it look to you? Or would you still want different function header on different platforms. (In that case, could you please provide some guidance on how to conditionally declare function header on different platforms?)